### PR TITLE
Client reconfiguration engine - complete implementation on replica side 

### PR DESCRIPTION
**Pull request by yontyon**
in this PR we complete the CRE implementation on the replica side.
1.  We maintain a state on the blockchain per CRE id (which is eventually a client node id)
2. We write and read the relevant data per CRE requests

To enable the above we made some more changes:
1. The sender ID was added to the reconfiguration request and to the validation method (in order to allow validation per client)
2. We allow self-verification in SigManager

To improve the CRE functionality each handler may return a WriteState which will be written to the blockchain. This WriteState also contains a callback that is invoked after the write command has succeeded. This allows us to perform actions write after updating the new state for the client (for example, after writing a new public key to the replica, rotate the private key).

Finally, we added a basic Apollo test that demonstrates the capabilities of CRE. By running the test, one can see in the logs a simulation of client key-exchange using CRE

To complete the implementation we need to decide how to verify the client reconfiguration requests. This can be either by using the transactions signing key or by having a dedicated key for CRE

### CODE REVIEW

File: reconfiguration/src/dispatcher.cpp [hunk 1]
```diff
@@ -52,7 +53,7 @@ ReconfigurationResponse Dispatcher::dispatch(const ReconfigurationRequest& reque
     // Run regular reconfiguration handlers
     for (auto& handler : reconfig_handlers_) {
       // Each reconfiguration handler handles only what it can validate
-      if (!handler->verifySignature(ser_data, ser_sig)) {
+      if (!handler->verifySignature(0, ser_data, ser_sig)) {
```
Comment (toly-kournik) [inline reconfiguration/src/dispatcher.cpp:56]: who's id is 0?

File: client/cre/src/client_reconfiguration_engine.cpp [hunk 2]
```diff
@@ -38,13 +38,13 @@ void ClientReconfigurationEngine::main() {
...
         }
-        State out_state;
+        WriteState out_state;
         if (!h->execute(update, out_state)) {
           LOG_ERROR(getLogger(), "error while executing the handlers");
           errored_handlers_.Get().Inc();
           continue;
         }
-        if (out_state.blockid == update.blockid) {
+        if (!out_state.data.empty()) {
```
Comment (toly-kournik) [inline client/cre/src/client_reconfiguration_engine.cpp:47]: if empty, does it means error?

File: kvbc/include/kvbc_key_types.hpp [hunk 3]
```diff
@@ -20,6 +20,13 @@ static const char reconfiguration_download_key = 0x26;
 static const char reconfiguration_install_key = 0x27;
 static const char reconfiguration_key_exchange = 0x28;
 static const char reconfiguration_add_remove = 0x29;
-static const char reconfiguration_client_key_exchange = 0x30;
+// 0x2a is reserved
```
Comment (toly-kournik) [inline kvbc/include/kvbc_key_types.hpp:23]: `static const char reconfiguration_reserved = 0x2a; //TODO [YB] check`

it shouldn't be reserved at a higher level

File: kvbc/include/kvbc_key_types.hpp [hunk 4]
```diff
@@ -20,6 +20,13 @@ static const char reconfiguration_download_key = 0x26;
...
+static const char reconfiguration_client_data_prefix = 0x2c;
 
+enum CLIENT_COMMAND_TYPES : uint8_t {
+  start_ = 0x0,
+  PUBLIC_KEY_EXCHANGE = 0x1,
```
Comment (toly-kournik) [inline kvbc/include/kvbc_key_types.hpp:28]: please put comments on every type

File: kvbc/include/reconfiguration_kvbc_handler.hpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+#pragma once
+
+#include "reconfiguration/ireconfiguration.hpp"
+#include "db_interfaces.h"
+#include "hex_tools.h"
+#include "block_metadata.hpp"
+#include "kvbc_key_types.hpp"
+#include "SigManager.hpp"
+#include "../../reconfiguration/include/reconfiguration/reconfiguration_handler.hpp"
```
Comment (toly-kournik) [inline kvbc/include/reconfiguration_kvbc_handler.hpp:21]: don't use relative path

File: kvbc/include/reconfiguration_kvbc_handler.hpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  bool handle(const concord::messages::ClientReconfigurationStateRequest&,
+              uint64_t,
+              concord::messages::ReconfigurationResponse&) override;
+
+ private:
+  concord::messages::ClientReconfigurationStateReply buildClientStateReply(
+      kvbc::BlockId, kvbc::keyTypes::CLIENT_COMMAND_TYPES command_type, uint32_t clientid);
+};
+/**
+ * TODO [YB] - add description
```
Comment (toly-kournik) [inline kvbc/include/reconfiguration_kvbc_handler.hpp:56]: What about description? ;-)

File: kvbc/include/reconfiguration_kvbc_handler.hpp [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  bool handle(const concord::messages::PruneRequest& command,
+              uint64_t sequence_number,
+              concord::messages::ReconfigurationResponse&) override;
+
+  bool handle(const concord::messages::ClientKeyExchangeCommand& command,
+              uint64_t sequence_number,
+              concord::messages::ReconfigurationResponse& response) override;
+};
+/**
+ * TODO [YB] - add description
```
Comment (toly-kournik) [inline kvbc/include/reconfiguration_kvbc_handler.hpp:104]: and here

File: tests/apollo/util/operator.py [hunk 8]
```diff
@@ -41,6 +41,7 @@ def _construct_reconfiguration_wedge_command(self):
             reconf_msg = cmf_msgs.ReconfigurationRequest()
             reconf_msg.command = wedge_cmd
             reconf_msg.additional_data = bytes(0)
+            reconf_msg.sender = 1000
```
Comment (toly-kournik) [inline tests/apollo/util/operator.py:44]: is it possible not to copy&paste the construction of ReconfigurationRequest?

File: tests/simpleKVBC/CMakeLists.txt [hunk 9]
```diff
@@ -1,6 +1,8 @@
 add_subdirectory(cmf)
 add_subdirectory(TesterClient)
 add_subdirectory(TesterReplica)
+add_subdirectory(TesterCRE)
```
Comment (toly-kournik) [inline tests/simpleKVBC/CMakeLists.txt:4]: Have you considered to include CRE into TesterCLient?

**Review by toly-kournik — APPROVED** [review]
Please address the comments.
LGTM!

File: client/cre/src/client_reconfiguration_engine.cpp (hunk 2 above)
Comment (yontyon) [reply client/cre/src/client_reconfiguration_engine.cpp:47]: no, it means that it doesn't want to update state on the blockchain for this action

**Pull request merged**